docs/content/policy-performance.md

```diff
@@ -131,6 +132,129 @@ For `glob.match(pattern, delimiter, match)` statements to be indexed the pattern
 | `glob.match("foo:**:bar", [":"], input.x)` | no | pattern contains `**` |
 | `glob.match("foo:*:bar", [":"], input.x[i])` | no | match contains variable(s) |
 
+### Comprehension Indexing
+
+Rego does not support mutation. As a result, certain operations like "group by" require
+use of comprehensions to aggregate values. To avoid O(n^2) runtime complexity in
+queries/rules that perform group-by, OPA may compute and memoize the entire collection
+produced by comprehensions at once. This ensures that runtime complexity is O(n) where
+n is the size of the collection that group-by/aggregation is being performed on.
+
+For example, suppose the policy must check if the number of ports exposed on an interface
+exceeds some threshold (e.g., any interface may expose up to 100 ports.) The policy is given
+the port->interface mapping as a JSON array under `input`:
+
+```json
+{
+  "exposed": [
+    {
+      "interface": "eth0",
+      "port": 8080,
+    },
+    {
+      "interface": "eth0",
+      "port": 8081,
+    },
+    {
+      "interface": "eth1",
+      "port": 443,
+    },
+    {
+      "interface": "lo1",
+      "port": 5000,
+    }
+  ]
+}
+```
+
+In this case, the policy must count the number of ports exposed on each interface. To do this,
+the policy must first aggregate/group the ports by the interface name. Conceptually,
+the policy should generate a document like this:
+
+```json
+{
+  "exposed_ports_by_interface": {
+    "eth0": [8080, 8081],
+    "eth1": [443],
+    "lo1": [5000]
+  }
+}
+```
+
+Since multiple ports could be exposed on a single interface, the policy must use a comprehension to
+aggregate the port values by the interface names. To implement this logic in Rego, we would write:
+
+```rego
+some i
+intf := input.exposed[i].interface
+ports := [port | some j; input.exposed[j].interface == intf; port := input.exposed[j].port]
+```
+
+Without comprehension indexing, this query would be O(n^2) where n is the size of `input.exposed`.
+However, with comprehension indexing, the query remains O(n) because OPA only computes the comprehension
+_once_. In this case, the comprehension is evaluated and all possible values of `ports` are computed
+at once. These values are indexed by the assignments of `intf`.
+
+To implement the policy above we could write:
+
+```rego
+deny[msg] {
+  some i
+  count(exposed_ports_by_interface[i]) > 100
+  msg := sprintf("interface '%v' exposes too many ports", [i])
+}
+
+exposed_ports_by_interface := {intf: ports |
+  some i
+  intf := input.exposed[i].interface
+  ports := [port |
+    some j
+    input.exposed[j].interface == intf
+    port := input.exposed[j].port
+  ]
+}
+```
+
+Comprehension indices can be built for any comprehension (nested or not) if the following restrictions are met:
+
```

Comment:
Would be helpful if there were a higher-level description of at least the common case.  Maybe something like "an iteration with a comprehension that is used to find a collection of objects that are equal to (or more generally compared to) the iteration variable".    Not happy with the wording, but something similar.

Perhaps also an example of what does NOT get indexed but might seem like it should.  Edit: you did this below.  Nice!

Reply:
Changed the description to this:

```

Indices can be built for comprehensions (nested or not) that generate collections (i.e., arrays, sets, or objects)
based on variables in an outer query. In the example above:

* `intf` is the variable in the outer query.
* `[port | some j; input.exposed[j].interface == intf; port := input.exposed[j].port]` is the comprehension.
* `ports` is the variable the collection is assigned to.
```

Then list out the conditions/exceptions.